index.ts

```diff
@@ -0,0 +1,552 @@
+  num: number;
+  abilities: SpeciesAbilities<AbilityT>;
+  baseForme?: SpeciesT;
+  baseSpecies?: SpeciesT;
+  baseStats: StatsTable;
+  battleOnly?: boolean;
+  cosmeticForms?: SpeciesT[];
+  doublesTier?: TierName;
+  eventOnly?: boolean;
+  evos?: SpeciesT[];
+  formeLetter?: string;
+  forme?: string;
+  isMega?: boolean;
+  isPrimal?: boolean;
+  isTotem?: boolean;
+  maleOnlyHidden?: boolean
+  maxHP?: number;
+  nfe: boolean;
+  otherFormes?: SpeciesT[];
+  prevo?: SpeciesT;
+  requiredAbility?: AbilityT;
+  requiredItems?: ItemT[];
+  requiredMove?: MoveT;
+  tier?: TierName;
+  types: TypeT[];
+  unreleasedHidden?: boolean
+  weightkg: number;
+}
+
+interface JSONSpecies extends JSONData & CommonSpecies<ID, ID, ID, ID, ID> {}
+interface Species extends Data & CommonSpecies<Ability, Item, Move, Species, Type> {
+  type 'Species';
+  cosmeticForms: Species[];
+  otherFormes: Speces[];
+  requiredItems: Item[];
+
+  toJSON(): JSONSpecies;
+}
+
+// TODO: name collision :(
+class Species {
+  getName(name: string): string;
+  getByID(id: ID): Species | undefined;
+  next(): {value: Species, done: boolean};
+}
+
+// species-details.ts
+
+// TODO: create EggGroup and Color enum types?
+type EvoType = 'trade' | 'stone' | 'levelMove' | 'levelExtra' | 'levelFriendship' | 'levelHold';
+
+interface CommonEventInfo<NatureT, AbilityT, MoveT> {
+  generation: Generation;
+  level?: number;
+  shiny?: 'always' | 'never' | 'possible';
+  gender?: GenderName;
+  nature?: NatureName;
+  ivs?: Partial<StatsTable>;
+  perfectIVs?: number;
+  isHidden?: boolean;
+  abilities?: AbilityT[];
+  maxEggMoves?: number;
+  moves?: MoveT[];
+  pokeball?: ID;
+  from?: string;
+}
+
+interface JSONEventInfo extends CommonEventInfo<NatureName, ID, ID> {}
+interface EventInfo extends CommonEventInfo<NatureName, Ability, Move> {}
+
+type MoveSource = string;
+
+interface CommonSpeciesDetails<EventInfoT, ItemT, MoveT> {
+  canHatch?: boolean;
+  color: string; // Color
+  eggGroups?: string[]; // EggGroup
+  encounters?: EventInfoT[]
+  eventPokemon?: EventInfoT[];
+  evoDetails?: {
+    type: EvoType;
+    condition?: string;
+    level?: number;
+    item?: ItemT;
+    move?: MoveT;
+  };
+  gender?: GenderName;
+  genderRatio?: {M: number, F: number};
+  heightm: number;
+  learnset: Map<ID, MoveSource[]>;
+}
+
+interface JSONSpeciesDetails extends CommonSpeciesDetails<JSONEventInfo, ID, ID> {}
+interface SpeciesDetails extends CommonSpeciesDetails<EventInfo, Item, Move> {
+  eggGroups: string[]; // EggGroup
+  encounters: EventInfoT[]
+  eventPokemon: EventInfoT[];
+
+  toJSON(): JSONSpeciesDetails;
+}
+
+// TODO: name collision :(
+class SpeciesDetails {
+  static forGen(gen: Generation, mod?: Mod): Promise<SpeciesDetails>;
+
+  getByID(id: ID): SpeciesDetails | undefined;
+}
+
+// stats.ts
+
+type StatName = 'hp' | 'atk' | 'def' | 'spa' | 'spd' | 'spe';
+type StatsTable<T = number> = {[stat in StatName]: T};
```

Comment:
`Map`

Reply:
I disagree - there are a defined number of keys

Reply:
> In short, you should always use a Map when you need a key-value collection. A good indicator that you need a collection is when you add and remove values dynamically from the collection, and especially when you don't know those values beforehand (e.g. they're read from a database, input by the user, etc).

> In contrast, you should be using objects when you know which and how many properties the object has while writing the code - when their shape is static. As Felix has put it: when you need a record. A good indicator for needing that is when the fields have different types, and when you never need to use bracket notation (or expect a limited set of property names in it).

https://stackoverflow.com/questions/32600157/maps-vs-objects-in-es6-when-to-use

Reply:
works for me